react-context/flow.txt

```diff
@@ -0,0 +1,67 @@
+User can sign in
+	- In index.js, the App react component is created with a constructor and rendered onto the browser page (client side)
+	- When App react component is rendered, the Navbar and PageContainer components are also rendered.
+	- After the first render, windoww.location.hash is #sign-in and the Auth component is rendered. The Auth component
+	describes the HTML for the sign-in page. **componenDidMount method is called after 1st render.
+	- When the Auth component is rendered, the AuthForm component is also rendered with key and action values as "sign-in" and an
+	"onSignIn" property with a callback function of handleSignIn (provided by AppContext provider in app.jsx)
+	- When the AuthForm component renders the sign-in form that includes the Username input, Password input, and "Register now"
+	message
+	- When the input value's change, the handleChange(event) method is called and a rerender is scheduled due to a statechange
+	- When the form is submited, the handleSubmit(event) method is called and a rerender is scheduled due to a statechange
+	- The handlesSubmit function call defines a req object with POST method, then a fetch request is made with the taarget as /api/autho/sign-in.
+	- The server receives the http request sent through fetch, and the app.post('api/auth/sign-in') function handles the
+	http request.
+	- The server send a request to the database with a sql query asking for the database row containing the username and password
+	specifide in the request body.
+	- If that record does not exist, the server sends a 401 error to the client. If the record exist, the argon2.verify method
+	returns whether the user's hashedPassword (stored in database) matches the provide password in the req.body.
+	- If the passwords match, a token is created and returned inside a json object along with the payload object (contains
+	userId, username) to the client.
+	- Any errors that occur are caught and handed over to express to handle.
+	- Back to the fetch request > once the res.json object is received from the server, if the user and token values exist
+	, the handleSignIn method of the App component class is called with the res.json object as the argument.
+	- When the handleSignIn method is called, the token is stored in localStorage and the state.user property is set
+	with the value of the user property of the res.json object
+	- Since there is a state change, the App component is rerendered
+	- Now we have a user value (from state), route: "sign-in", isAuthorizing: false
+	- Since there is a user value now, when the Auth component is rendering (called from renderPage()), the Redirect component gets rendered.
+	- The Redirect component sets the window.location.hash to ''
+	- ???? // not sure how and when the route state get's set to '', but when it does, a rerender occurs
+	- When the App component's render function is called, the renderPage() function gets called and returns the Home component which
```

Comment:
The redirect triggers a `'hashchange'` event by assigning a value to `window.location.hash`